### PR TITLE
sysupdated: Bugfixes & improvements

### PR DESCRIPTION
**Pull request by AdrianVovk**
A relatively major bugfix for sysupdated, then some minor follow-up improvements. See commit messages

### CODE REVIEW

File: src/sysupdate/sysupdate-transfer.c [hunk 1]
```diff
@@ -957,17 +964,32 @@ static int helper_on_notify(sd_event_source *s, int fd, uint32_t revents, void *
...
-        if (progress < 0) {
-                log_warning("Got invalid percent value '%s', ignoring.", progress_str);
-                return 0;
+        if (errno_str) {
+                truncate_nl(errno_str);
+                r = parse_errno(errno_str);
+                if (r < 0)
+                        log_warning_errno(r, "Got invalid errno value '%s', ignoring: %m", errno_str);
+                else {
+                        ctx->helper_errno = r;
```
Comment (yuwata) [inline src/sysupdate/sysupdate-transfer.c:975]: Here, positive value is assigned.

File: src/sysupdate/sysupdate-transfer.c [hunk 2]
```diff
@@ -886,29 +887,34 @@ static int callout_context_new(const Transfer *t, const Instance *i, TransferPro
...
                 if (IN_SET(si->si_code, CLD_KILLED, CLD_DUMPED))
                         log_error("%s terminated by signal %s.", ctx->name, signal_to_string(si->si_status));
                 else
                         log_error("%s failed due to unknown reason.", ctx->name);
         }
 
-        pidref_done(&ctx->pid);
-
-        return sd_event_exit(sd_event_source_get_event(s), code);
+        return sd_event_exit(sd_event_source_get_event(s), r);
```
Comment (yuwata) [inline src/sysupdate/sysupdate-transfer.c:917]: Then, the positive value is still propagated??

File: src/sysupdate/sysupdated.c [hunk 3]
```diff
@@ -257,9 +257,7 @@ static void job_on_ready(Job *j) {
                 log_warning_errno(r, "Failed to run callback on job ready event, ignoring: %m");
 }
 
-static void job_on_errno(Job *j, char *b) {
-        /* Take ownership of donated buffer */
-        _cleanup_free_ char *buf = TAKE_PTR(b);
+static void job_on_errno(Job *j, char *buf) {
```
Comment (yuwata) [inline src/sysupdate/sysupdated.c]: When you touch this, please change `assert_se(buf)` below to `assert(buf)`.

File: src/sysupdate/sysupdated.c [hunk 4]
```diff
@@ -303,8 +299,8 @@ static void job_on_version(Job *j, char *version) {
         assert(j);
         assert_se(version);
 
-        /* Take ownership of donated memory */
-        free_and_replace(j->version, version);
+        if (strdup_to(&j->version, version) < 0)
+                return (void) log_oom();
```
Comment (yuwata) [inline src/sysupdate/sysupdated.c]: `free_and_strdup_warn()`

File: src/sysupdate/sysupdated.c [hunk 5]
```diff
@@ -335,7 +337,7 @@ static int job_on_exit(sd_event_source *s, const siginfo_t *si, void *userdata)
         else {
                 r = job_parse_child_output(TAKE_FD(j->stdout_fd), &json);
                 if (r < 0)
-                        sd_bus_error_set_errnof(&error, r, "Failed to parse JSON: %m");
+                        sd_bus_error_set_errnof(&error, r, "Failed to parse job worker output: %m");
```
Comment (yuwata) [inline src/sysupdate/sysupdated.c:341]: missing job id here?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/sysupdate/sysupdate-transfer.c (hunk 2 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdate-transfer.c:917]: Heh whoops. This is what I get for opening PRs after midnight on a Friday :upside_down_face:

File: src/sysupdate/sysupdated.c (hunk 5 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdated.c:341]: That's intentional.

The dbus client that this error delivered to doesn't need a job ID. Actually it'll probably have a better way to log about failure. For instance, updatectl will include the name of the target that failed which is a lot more actionable to the user on the CLI than a job ID.

And on the journal, where the job ID is useful, it'll be included when this error is printed out (a couple of lines below: `Job <id> failed with bus error, ignoring callback: <bus error message>`)

**Review by yuwata — APPROVED** [review]

File: src/sysupdate/sysupdated.c [hunk 6]
```diff
@@ -302,10 +298,10 @@ static void job_on_progress(Job *j, char *b) {
 
 static void job_on_version(Job *j, char *version) {
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: should gain a "const" now

File: src/sysupdate/sysupdated.c [hunk 7]
```diff
@@ -277,14 +275,12 @@ static void job_on_errno(Job *j, char *b) {
         log_debug_errno(r, "Got errno from job %" PRIu64 ": %i (%m)", j->id, r);
 }
 
-static void job_on_progress(Job *j, char *b) {
-        /* Take ownership of donated buffer */
-        _cleanup_free_ char *buf = TAKE_PTR(b);
+static void job_on_progress(Job *j, char *buf) {
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: 2nd param should gain a "const" now

File: src/sysupdate/sysupdated.c [hunk 8]
```diff
@@ -258,13 +258,11 @@ static void job_on_ready(Job *j) {
                 log_warning_errno(r, "Failed to run callback on job ready event, ignoring: %m");
 }
 
-static void job_on_errno(Job *j, char *b) {
-        /* Take ownership of donated buffer */
-        _cleanup_free_ char *buf = TAKE_PTR(b);
+static void job_on_errno(Job *j, char *buf) {
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: 2nd arg should gain a const now

**Comment by AdrianVovk** [comment]
CI failures appear unrelated:

`mkosi / ci (centos, 10, 0, -Og, yes) (pull_request)`:

```
No match for argument: python3dist(pillow)
Error: Unable to find a match: python3dist(pillow)
```

`testing-farm:fedora-rawhide-x86_64`:

```
TEST-69-SHUTDOWN:                   FAIL     (  8 s)
<snip>
TOTAL FAILURES: 1 OF 64
```

`nobile-amd64`:

```

11211s Summary of Failures:
11211s 
11211s  2/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-iscsi_lvm                                FAIL              9.49s   (exit status 247 or signal 119 SIGinvalid)
11211s 
11211s Ok:                 89  
11211s Expected Fail:      0   
11211s Fail:               1   
11211s Unexpected Pass:    0   
11211s Skipped:            2   
11211s Timeout:            0   
```

**Pull request merged**